Costomize the sleep test

touilkhouloud opened the pull request:
Costomize the sleep test to use parameters and run the suited tests.
The parameters to be added in the test-config.yaml file to each board
that supports the test.
Add a check for the existence of the rtc otherwise it won't execute
the test.

Signed-off-by: Khouloud Touil <ktouil@baylibre.com>

khilman commented on templates/sleep/sleep.sh:
```diff
@@ -0,0 +1,30 @@
+		fi
+	done
+
+	if echo "${test_params}" | grep -q "freez" ; then
```
s/freez/freeze/

gctucker commented on test-configs.yaml:
```diff
@@ -210,6 +210,8 @@ device_types:
       # added to multi_v7_defconfig as of 5.0.
       - blocklist: {kernel: ['v3.', 'v4.']}
+    params:
+      test_params: 'freez'
```
The `params` used in device types should only be things that apply to the device.  The parameters you're adding are for the sleep test plan, so they should not be put here.

A better way to do that would be to create variants of the `sleep` test plan with the parameters, and then update the `test_config` entries to use those variants on the relevant device types.

```yaml
  sleep:
    rootfs: debian_buster_ramdisk
    params:
        sleep_params: standby freeze mem

  sleep_mem:
    base_name: sleep
    rootfs: debian_buster_ramdisk
    params:
        sleep_params: mem
```
Then for devices that should only be doing `mem`, picking a random one:
```yaml
  - device_type: hifive-unleashed-a00
    test_plans:
      - baseline
      - sleep_mem
```
Looking at the devices where you added `freeze`, that matches with all the test configs enabling `sleep` so `sleep_mem` isn't actually needed in practice unless we decide that some devices shouldn't be tested with other modes.

touilkhouloud replied:
I asked @khilman about the boards to support freeze/mem, he told me to make mem a default test and all what we have now supports freeze that's why in the sleep.sh I didn't check for a param mem only for freeze.

gctucker replied:
Ah, that's slightly different then.  It means all the boards that don't currently run the "sleep" test plan can start running it with only "mem" (suspend to RAM) since it should be supported on all systems.  Support for "freeze" may be broken on some platforms.

So following the YAML example above, it would mean leaving all the test configs with "sleep" as they currently are, and adding "sleep_mem" to other devices to run just the "mem" test.

touilkhouloud replied:
I am not sure about the other boards, I tried one and it doesn't have RTC at all, what I suggest is to create a rtcwake test plan that have freeze and mem as params for the boards to run sleep test plan now and create rtcwake-mem to be used in the next future for boards just how support mem such as meson sei610.
And I am not sure about standby because we don't test it for now, do you want to add it ? and do you have any idea about the boards to support it ?

gctucker replied:
If you add it to the test plan in this PR, it will get tested on staging.

The first version of the "sleep" plan was a very basic one, I believe we should be trying all the suspend modes and only skip those that are known to not be supported on some platforms.

gctucker reviewed (changes requested):
A few more things to tweak, this should actually simplify the code quite a bit.

About using test definitions from a git repository, this is introducing the issue of versioning.  The safe way to do this right now is to use the `kernelci.org` branch which gets updated in production at the same time as the rest of the configuration and the code.

touilkhouloud commented on templates/sleep/sleep.sh:
```diff
@@ -0,0 +1,30 @@
+		done
+	fi
+else
+	lava-test-case rtc-exist --result fail
```
I agree with Guillaume, as we are precising in the test-config.yaml file which file should run a sleep test so we know that specific board support RTC so if it doesn't exist it should fail and investigate why it's failing.

touilkhouloud commented on templates/sleep/sleep.jinja2:
```diff
@@ -2,28 +2,11 @@
       name: sleep
-      path: inline/sleep.yaml
+      parameters:
+        test_params: "{{ test_params }}"
```
sure, sleep_params then !

touilkhouloud commented on templates/sleep/sleep.jinja2:
```diff
@@ -2,28 +2,11 @@
-      lava-signal: kmsg
-      from: inline
+    - repository: https://github.com/kernelci/kernelci-core
+      branch: master
```
Sure !

touilkhouloud commented on templates/sleep/sleep.sh:
```diff
@@ -0,0 +1,30 @@
+#!/bin/sh
+
+test_params=$1
```
okay I will do it this way then.

gctucker commented:
One more thing, should we rename this test plan `rtcwake` instead of `sleep`?  I've always find the name a bit ambiguous, as it could just mean that a process is sleeping, or a CPU is idle...  Half of the test is checking that the system can resume so it's not just "sleeping" but also "waking".

gctucker commented:
Skipping on staging until the comments have been addressed, to avoid merge conflicts with other PRs (kselftest in particular).

touilkhouloud commented:
Hello @gctucker @khilman , so I modified the changes as you suggested,  I added a new test plan sleep-mem as mentioned, but it came to my mind that the purpose of this is to have one test plan and using the params we can execute the wanted part but if we are going to create a test plan for each different mode that won't need the params anymore and it's not the purpose of the customization, you can see the second commit, waiting for your reviews.

When we finish the structure of the code, I will add anther commit for renaming sleep, this will be the last part to be handled.

gctucker commented:
Sorry if my comments weren't clear, the idea is not to add a new test plan but define variants in `test-configs.yaml` with parameters.

touilkhouloud commented:
Pushed again, I hope I got what you mean this time ;) !

gctucker commented:
Thanks, yes I think that's great :)

Now there might be a merge conflict with the kselftest PRs, so I'll leave the `staging-skip` label today but this should get tested on staging next week.

khilman reviewed (approved):
Looks good to me.  Thanks!

gctucker commented:
Removing the `staging-skip` tag as this isn't conflicting with other PRs any more.

gctucker commented:
I've added a patch on staging to get the test definitions from the `staging.kernelci.org` branch, will see if we now get some results.

gctucker commented:
OK here are some results: https://staging.kernelci.org/test/job/kernelci/branch/staging.kernelci.org/kernel/staging-20200917.1/plan/sleep/

@touilkhouloud @khilman By the way, maybe you would like to run some `sleep` tests in lab-baylibre? (worth a separate PR though)

gctucker commented:
It looks like even devices that suspend/resume correctly are failing with the "standby" mode:
https://staging.kernelci.org/test/plan/id/5f6322c0e125fb11ab10f3d4/

So I wonder if it's worth enabling it by default, prior to this PR it was never tested.  I'm not entirely sure if it's expected to pass, would need to see what rtcwake actually does in this mode.

Other than that this PR seems good to go now.

khilman commented:
Correct, we should drop the standby tests since the weren't enabled
before.  I'm not familiar with standby, but I think it's mostly an
x86 thing, so should be added separately if desired.

touilkhouloud commented:
Hello @gctucker,  so should I change this PR by removing the standby mode ?

gctucker commented on test-configs.yaml:
```diff
@@ -167,6 +167,15 @@ test_plans:
   sleep:
     rootfs: debian_buster_ramdisk
+    params:
+      sleep_params: mem standby freeze
```
```suggestion
      sleep_params: mem freeze
```

gctucker reviewed (commented):
@touilkhouloud Yes, as in the suggestion above.

gctucker commented:
Tested OK on staging:
https://staging.kernelci.org/test/job/kernelci/branch/staging.kernelci.org/kernel/staging-20200923.0/plan/sleep/

On a side note: There is one inherent issue with suspend/resume tests, when resume completely fails and the device is stuck the failure is not reported.  That's something which might be improved with a timeout set on each test case, but I'm not sure LAVA can do that easily.  We could also have a KernelCI feature to provide a list of required test cases to be run in the meta-data, and those missing would be considered as failures.
